### ISSUE TITLE
Deprecate adding a (meaningless) colorbar to RGB or RGBA images

### ISSUE THREAD

**Issue by Zac-HD** [matplotlib/matplotlib#20576]
### Describe the issue

In #9391, I proposed to [open a new issue for "colorbar should refuse to let itself be created for an RGB(A) image"](https://github.com/matplotlib/matplotlib/issues/9391#issuecomment-356577162) because colorbars are not meaningful for images with existing color.  https://github.com/matplotlib/matplotlib/issues/4369#issuecomment-95383235 similarly notes that this makes no sense.

(better late than never I guess!)

### Proposed fix

I think that attempting to add a colorbar for an RGB or RGBA image should be an explicit error, following a period of deprecation warnings.

**Comment by jklymak** [comment]
A colorbar is just an artist.  I'm not aware of us being prescriptive about the use of _any_ artist just because it is not sensible.  

I suppose an argument can be made that a colormapable that does not use a norm should make an empty colorbar; we don't currently track on the mappable if it ever did a mapping or not.

**Comment by timhoffm** [comment]
One could consider it a design flaw that `AxesImage` inherits from `ScalarMappable`, but that ship has sailed.

**Comment by jklymak** [comment]
Well, a colorbar means something for `imshow` in general, just not when `imshow` is called with with RGBA...

**Comment by jkfindeisen** [comment]
For example, the colorbar could re-create exactly the colormap of the used image. It might still be desired to show one, for example to resemble the layout of other plots.

**Comment by timhoffm** [comment]
> For example, the colorbar could re-create exactly the colormap of the used image.

For a RGB image there is no reasonable way to map the values to any linear range (and there's also no scale). The only reasonable colorbar is an empty colorbar.

**Comment by jklymak** [comment]
I don't think any of this is hard to do (check mappable is AxesImage, and don't fill the colorbar if `get_array` is 3- or 4-D, or emit a warning/error).  But I'm still not sure why this is a guard rail we need to put up, or why putting the guardrail up would be less confusing than the status quo.

**Comment by timhoffm** [comment]
Taking the original example code of #9391 we get:
![grafik](https://user-images.githubusercontent.com/2836374/124654101-6a076280-de9e-11eb-9d9b-373a77f22981.png)

The RGB cases are clearly wrong: There is a colorbar implying a colormapping, but obviously the mapping is not related to the image. Even worse the scale seems to be influenced by the data. IMHO this falls under the case "errors should never pass silently".

**Comment by jklymak** [comment]
OK, in that case I guess we should error if `mappable.get_array().dim > 2` and just say that such a mappable cannot have a colorbar.  Right now if we pass a Line2D we get
```
AttributeError: 'Line2D' object has no attribute 'get_array'
```
which isn't the greatest error message either.

**Comment by jklymak** [comment]
Maybe:
```
       if mappable is None:
            mappable = cm.ScalarMappable(norm=norm, cmap=cmap)
        elif not hasattr(mappable, 'get_array'):
            raise ValueError('colorbar mappable must be a subclass of '
                             'ScalarMappable or have a get_array method.')
        elif mappable.get_array().ndim > 2:
            raise ValueError('Mappable array has more than '
                             'two dimensions, and hence cannot be mapped '
                             'using a colorbar; perhaps because you '
                             'passed RGB(A) data to the method that '
                             'made the mappable.')
```

**Comment by timhoffm** [comment]
The Line2D case is slightly different. Due to duck typing we don't thoroughly check the input types. If it's wrong it blows up somehow. But passing Line2D is a clear violation of the API, which states that you have to pass a ScalarMappable.

The RGB AxesImage is different in that it *is* a ScalarMappable and formally fulfills the API contract. It's only that it cannot be mapped to colors (which as written above is a design flaw). So we need to be more comforting to the user and give a nicer error.

**Comment by jklymak** [comment]
Then I guess the remaining problem is what if someone "needs" to add an empty colorbar.  We don't currently have an easy way to do that without hoisting a fake ScalarMappable.

**Comment by timhoffm** [comment]
> Then I guess the remaining problem is what if someone "needs" to add an empty colorbar. We don't currently have an easy way to do that without hoisting a fake ScalarMappable.

I question the "need". I claim that an empty colorbar is not what users *really* want, and at least nobody has asked for it so far. What is the purpose of an empty colorbar? It's just visual noise. It would almost always be better to keep an empty space instead of the colorbar (that's in general also not simple, but another topic).

If empty colorbar is really a thing, let's wait until user demand comes up. We can then always add docs how to create one or even add helper code.

**Comment by jklymak** [comment]
Anyone is free to take up the above patch.  I don't really feel strongly enough to add this to the code base, but if others feel strongly about it, they should go ahead....

**Comment by jkfindeisen** [comment]
> For a RGB image there is no reasonable way to map the values to a linear range (and there's also no scale). The only reasonable colorbar is an empty colorbar.

I can easily create an RGB image from a data array and a colormap, then store the RGB image, terminate the program, load the RGB image again, display it and display the same colormap and it would be a reasonable output. One could even try to reverse engineer a data array from an RGB image given a colormap but that's is outside of the scope here. Here, I guess, it's more about what is reasonable and meaningful and I can only say that I had such cases and displayed RGB images with a colorbar (in Matlab) and it was a meaningful outcome to me because I knew what I was doing.

I guess the idea is to prevent people from doing things that likely aren't meaningful, but it also restricts the freedom of what you can do with the software. Maybe a warning ("this may not be meaningful") would be a better compromise.

But I guess (without inside knowledge) that if one really wants, one could probably create a colorbar object not attached to any image and simply position it where the colorbar would normally be, so allowing colorbars for RGB images isn't really limiting anyone.

**Comment by timhoffm** [comment]
Yes, you *can* create arbitrary colorbars that are not related to an image. In that case, you have to state which value range, and how the mapping is done (via norm and colormap). And I think this is how it should be. So we are not limiting any use case (including supporting your example), but you have to be explicit what you want.

**Comment by jkfindeisen** [comment]
Then I think that it's fine.

**Comment by Zac-HD** [comment]
I would still like this feature.

(and I'm not a fan of closing issues based on age alone)

**Comment by jklymak** [comment]
An attempt was made in [20630](https://github.com/matplotlib/matplotlib/pull/20630), but foundered on all the different cases.

Perhaps you could explain how this would practically trip someone up?  It seems pretty straightforward to keep track of whether you are mapping data or just displaying an image.  

> (and I'm not a fan of closing issues based on age alone)

The bot brings things up for discussion again, versus leaving to the dustbin of history.  Whether something is closed or not depends on the outcome of that discussion.   

However, I also think a stronger argument needs to be made that this is a practical problem that is worth the code complexity to fix.

**Comment by tacaswell** [comment]
My read of the current state is:

 - putting a colorbar on an RGB(A) image is in almost all (but not _all_) cases wrong [I think this is non-controversial]
 - we don't have a good way to tell if an AxesImage is in colormapping or RGB(A) mode (due to the mixing with ScalarMappable (and soon VectorMappable) there are ambiguities) [again I think this is not controversial]
 - even if we solved the previous problem, we can change which mode the AxesImage is in at run time!


```python
import matplotlib.pyplot as plt
import numpy as np
plt.ion()
im = plt.imshow(np.random.rand(5, 7))
# This is OK
plt.colorbar()
# and now the colorbar is meaningless!
im.set_data(np.random.rand(5, 7, 3))
```


I don't think we want to go so far as to remove a colorbar of the image it is linked to has RGB(A) set to it, so I don't think we want to error if we are asked to create one. 

However I think a warning (special cased to the Image classes) in the colorbar code is the most pragmatic option.  That gets us out of solving the general problem and does not open up path-dependency on getting to an not-meaningful state, but does let users know that they have done something we are pretty sure is a mistake.  The user can then either notice that they made a mistake (which saves the user from a foot-cannon) or can switch to the "dummy" ColorMappable mentioned above to silence the error (where the user has confirmed they definitely want to use the foot-cannon).  

------

@Zac-HD I'm 100% with you on not auto-closing on age but we have found this process (flag old issues, give ourselves 30days to respond at all) is pretty good at re-surfacing issues with value.
